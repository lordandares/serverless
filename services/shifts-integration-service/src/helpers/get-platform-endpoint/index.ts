interface DeviceReceiver {
  deviceToken: string
  platform: string
}

export default async function getPlatformEndpoint(
  client: AWS.SNS,
  receiver: DeviceReceiver,
): Promise<AWS.SNS.CreateEndpointResponse | void> {
  const platformConfig = {
    android: {
      arn: process.env.AWS_SNS_ANDROID_ARN,
      service: process.env.AWS_SNS_ANDROID_SERVICE,
    },
    ios: {
      arn: process.env.AWS_SNS_IOS_ARN,
      service: process.env.AWS_SNS_IOS_SERVICE,
    },
  }

  const { deviceToken, platform } = receiver

  if (!deviceToken || !platform) return

  const config = platformConfig[platform]

  const params = {
    PlatformApplicationArn: config.arn,
    Token: deviceToken,
  }

  const endpoint = await client.createPlatformEndpoint(params).promise()

  return endpoint
}
