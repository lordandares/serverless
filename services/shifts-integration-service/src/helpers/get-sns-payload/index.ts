export default function getSnsPayload(params) {
  const { endpointArn, message, title, type: notificationType } = params

  return {
    Message: JSON.stringify({
      default: message,
      APNS: JSON.stringify({
        aps: {
          alert: message,
          'content-available': 1,
          sound: 'default',
        },
        notificationType,
      }),
      APNS_SANDBOX: JSON.stringify({
        aps: {
          alert: message,
          'content-available': 1,
          sound: 'default',
        },
        notificationType,
      }),
      GCM: JSON.stringify({
        data: {
          data: {
            notificationType,
          },
          message,
          title,
        },
      }),
    }),
    MessageStructure: 'json',
    TargetArn: endpointArn,
  }
}
