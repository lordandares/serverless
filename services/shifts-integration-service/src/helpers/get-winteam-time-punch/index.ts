import { errors, mongo, schemas } from '@lighthouse/serverless-common'
import { getOr, last } from 'lodash/fp'
import * as moment from 'moment-timezone'
import * as querystring from 'querystring'

// 2018-08-03 16:00:00.000-0400
const DATE_FORMAT = 'YYYY-MM-DD HH:mm:ss.SSSZZ'

interface TimePunchEvent extends TimePunch {
  event: string
}

interface TimePunch {
  result: {
    CallLogId: string
    HoursLogID?: string
    JobDescription: string
    JobID?: string
    JobNumber?: string
    PunchStatus: string
    PunchTime: string
    StatusReason?: string
  }
  shift: schemas.ShiftSchema
  user: schemas.UserSchema
}

interface TimePunchResponse {
  GeoCoordinate?: {
    Latitude: number
    Longitude: number
  }
  EmployeeId?: string
  EmployeeNumber?: string
  JobId?: string
  JobNumber?: string
  PunchTime: string
}

export async function getWinTeamTimePunch(
  timePunchEvent: TimePunchEvent,
): Promise<TimePunchResponse | void> {
  const { event, shift, user } = timePunchEvent
  const { application: applicationId, location: locationId } = shift

  if (!event) {
    throw new errors.ValidationError({
      message: 'Missing event from TimePunch',
    })
  }

  const employeeReference = getEmployeeReference(user)

  if (!employeeReference) {
    console.info('GetWinTeamTimePunch :: Missing Employee Reference', {
      shift,
      user,
    })

    return
  }

  const shiftEntity = getShiftEntity(shift, event)
  const shiftTimestamp = getOr(false, 'time', shiftEntity)

  if (!shiftTimestamp) {
    throw new errors.ValidationError({
      message: 'Missing timestamp from TimePunch',
      data: {
        shiftEntity,
        shiftTimestamp,
      },
    })
  }

  const jobReference = await getJobReference(
    shiftEntity,
    applicationId,
    locationId,
  )

  const punchTime = moment.utc(shiftTimestamp).format(DATE_FORMAT)

  const payload: TimePunchResponse = {
    ...employeeReference,
    ...jobReference,
    PunchTime: punchTime,
  }

  // Append GeoCoordinates if they exist
  const gpsCoordinates = getOr([], 'gps.geometry.coordinates', shiftEntity)

  if (gpsCoordinates && gpsCoordinates.length === 2) {
    payload.GeoCoordinate = {
      Latitude: gpsCoordinates[1],
      Longitude: gpsCoordinates[0],
    }
  }

  return payload
}

export async function getWinTeamMessages(
  timePunch: TimePunch,
): Promise<string | undefined> {
  const { result, shift, user } = timePunch

  const { JobNumber } = result

  const employeeReference = getEmployeeReference(user)
  if (!employeeReference) {
    console.info('GetWinTeamMessages :: Missing Employee Reference', {
      shift,
      user,
    })

    return
  }

  if (!JobNumber) {
    console.info('GetWinTeamMessages :: Missing Job Reference', {
      result,
      shift,
      user,
    })

    return
  }

  const punchTime = moment.utc(shift.start.time).format(DATE_FORMAT)

  const params = {
    ...employeeReference,
    JobNumber,
    PunchTime: punchTime,
  }

  return querystring.stringify(params)
}

export function getEmployeeReference(user: schemas.UserSchema): object | void {
  const employeeId = getOr(false, 'plugins.winteam.options.employeeId', user)
  const employeeNumber = getOr(
    false,
    'plugins.winteam.options.employeeNumber',
    user,
  )

  if (employeeId || employeeNumber) {
    return { EmployeeNumber: employeeId || employeeNumber }
  }
}

export async function getJobReference(
  shiftEntity: schemas.ShiftEventSchema,
  applicationId: string,
  locationId?: string,
): Promise<object | void> {
  const areaLocationId = getOr(false, 'area.location.id', shiftEntity)

  // If within an area location / legacy location, attempt to find the job number
  if (areaLocationId || locationId) {
    const isLegacy = !areaLocationId

    const collectionName = isLegacy ? 'locations' : 'areas'

    const queryLocationId = isLegacy ? locationId : areaLocationId
    const query = {
      _id: new mongo.ObjectId(queryLocationId),
      application: new mongo.ObjectId(applicationId),
    }

    const collection = await mongo.getCollection(collectionName)
    const location = await collection.findOne(query)

    if (location) {
      const jobId = getOr(false, 'plugins.winteam.options.jobId', location)
      const jobNumber = getOr(
        false,
        'plugins.winteam.options.jobNumber',
        location,
      )

      if (jobId || jobNumber) {
        return { JobNumber: jobId || jobNumber }
      }
    }
  }
}

export function getShiftEntity(
  shift: schemas.ShiftSchema,
  event: string,
): object | void {
  switch (event) {
    case 'shift-break-end':
      return last<schemas.ShiftBreakSchema>(shift.breaks)!.end
    case 'shift-end':
      return shift.end
    case 'shift-break-start':
      return last<schemas.ShiftBreakSchema>(shift.breaks)!.start
    case 'shift-start':
      return shift.start
    default:
      return {}
  }
}
