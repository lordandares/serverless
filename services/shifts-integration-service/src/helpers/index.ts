export { default as getPlatformEndpoint } from './get-platform-endpoint'
export { default as getSnsPayload } from './get-sns-payload'
export {
  getWinTeamMessages,
  getWinTeamTimePunch,
} from './get-winteam-time-punch'
export { default as notifyUser } from './notify-user'
export { default as parseWinTeamMessages } from './parse-winteam-messages'
export { default as sendMessage } from './send-message'
export {
  default as validateEndpointAttributes,
} from './validate-endpoint-attributes'
