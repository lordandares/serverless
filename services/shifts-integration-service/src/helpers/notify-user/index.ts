import { errors, schemas } from '@lighthouse/serverless-common'
import * as AWS from 'aws-sdk'
import { getOr, isEmpty } from 'lodash/fp'

import { getPlatformEndpoint, sendMessage } from '../'

interface NotifyUserParams {
  message: string
  user: schemas.UserSchema
}

export default async function notifyUser(
  params: NotifyUserParams,
): Promise<void> {
  const { message, user } = params

  if (!message || !user) {
    throw new errors.ValidationError({
      data: {
        message,
        user,
      },
      message: 'notifyUser :: Missing required params',
    })
  }

  const deviceReceivers = getOr([], 'speakerbox.deviceReceivers', user)

  if (isEmpty(deviceReceivers)) return

  const sns = new AWS.SNS()

  const endpoints: AWS.SNS.CreateEndpointResponse[] = await Promise.all(
    deviceReceivers.map(
      async receiver => await getPlatformEndpoint(sns, receiver),
    ),
  )

  await sendMessage(sns, endpoints, params)
}
