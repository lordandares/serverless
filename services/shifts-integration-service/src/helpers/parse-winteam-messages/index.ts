import { mongo } from '@lighthouse/serverless-common'
import { getOr } from 'lodash/fp'

export default function parseWinTeamMessages(winteamResponse) {
  const messages = getOr([], 'Messages', winteamResponse)

  return messages.map(message => {
    const {
      _id = new mongo.ObjectId(),
      Id,
      IsGlobal,
      Message,
      Response1,
      Response2,
      Response1Id,
      Response2Id,
      PunchActionID,
    } = message

    return {
      _id,
      isGlobal: !!IsGlobal,
      message: Message,
      messageId: Id,
      punchActionId: PunchActionID,
      response1: Response1,
      response1Id: Response1Id,
      response2: Response2,
      response2Id: Response2Id,
    }
  })
}
