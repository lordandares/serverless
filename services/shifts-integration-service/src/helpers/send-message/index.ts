import { getSnsPayload, validateEndpointAttributes } from '../'

interface SendMessageParams {
  message: string
  title: string
  type: string
}

export default async function sendMessage(
  client: AWS.SNS,
  endpoints: AWS.SNS.CreateEndpointResponse[],
  params: SendMessageParams,
): Promise<object[]> {
  const result = await Promise.all(
    endpoints.map(async endpoint => {
      const { EndpointArn: endpointArn } = endpoint

      await validateEndpointAttributes(client, endpointArn)

      const payload = getSnsPayload({ endpointArn, ...params })

      console.info('sending message', params)

      const result = await client.publish(payload).promise()

      console.info('sent message', result)
      return result
    }),
  )

  return result
}
