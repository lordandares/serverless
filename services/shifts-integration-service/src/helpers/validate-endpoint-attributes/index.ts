export default async function validateEndpointAttributes(
  client: AWS.SNS,
  arn: string,
): Promise<void> {
  const { Attributes } = await client
    .getEndpointAttributes({
      EndpointArn: arn,
    })
    .promise()

  const isEnabled = Attributes && Attributes.Enabled === 'true'

  if (isEnabled) return

  await client
    .setEndpointAttributes({
      EndpointArn: arn,
      Attributes: {
        ...Attributes,
        Enabled: 'true',
      },
    })
    .promise()
}
