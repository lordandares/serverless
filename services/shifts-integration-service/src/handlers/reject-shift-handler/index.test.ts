jest.mock('../../helpers')

import * as MockDate from 'mockdate'
import { mongo } from '@lighthouse/serverless-common'

mongo.getCollection = jest.fn()

import rejectShiftHandler, { formatShiftErrors } from './'
import { notifyUser } from '../../helpers'

describe('handlers:reject-shift-handler', () => {
  beforeEach(() => jest.resetAllMocks())

  it('should error if the `shiftId` is missing', async () => {
    expect.assertions(1)

    const event = {
      shift: {
        _id: '',
      },
    }

    expect(rejectShiftHandler(event)).rejects.toThrowErrorMatchingSnapshot()
  })

  it('should end a shift and set the status to rejected', async () => {
    expect.assertions(6)

    notifyUser.mockResolvedValue({})

    const expectedEnd = '2018-06-15T14:00:00.000Z'
    MockDate.set(expectedEnd)

    const event = {
      errors: [
        {
          AttemptedValue: 'invalid@email',
          ErrorMessage: 'Invalid Email Address',
          ErrorType: 'ValidationError',
          FieldName: 'email',
        },
      ],
      shift: {
        _id: '5d1eb8a4fc13ae1f85000000',
      },
      user: {
        _id: '5d1eb8a4fc13ae1f85000001',
      },
    }

    const mockFindOneAndUpdate = jest.fn().mockResolvedValue({
      value: {
        ...event.shift,
        end: {
          time: new Date(expectedEnd),
        },
        status: 'rejected',
        user: '1234',
      },
    })

    mongo.getCollection.mockResolvedValue({
      findOneAndUpdate: mockFindOneAndUpdate,
    })

    const result = await rejectShiftHandler(event)

    expect(mongo.getCollection).toHaveBeenCalledTimes(1)
    expect(mongo.getCollection).toBeCalledWith('shifts')

    expect(mockFindOneAndUpdate).toHaveBeenCalledTimes(1)
    expect(mockFindOneAndUpdate).toBeCalledWith(
      { _id: new mongo.ObjectId(event.shift._id) },
      {
        $set: {
          end: {
            time: new Date(expectedEnd),
          },
          status: 'rejected',
          verificationErrors: [
            {
              attemptedValue: 'invalid@email',
              errorMessage: 'Invalid Email Address',
              errorType: 'ValidationError',
              field: 'email',
            },
          ],
        },
      },
      { returnOriginal: false },
    )

    expect(notifyUser).toHaveBeenCalledTimes(1)

    expect(result).toMatchSnapshot()
  })

  describe('helpers :: formatShiftErrors', () => {
    it('should return an empty array if no errors are present', () => {
      const errors = []

      expect(formatShiftErrors(errors)).toEqual([])
    })

    it('should format error messages', () => {
      const errors = [
        {
          AttemptedValue: 'invalid@email',
          ErrorMessage: 'Invalid Email Address',
          ErrorType: 'ValidationError',
          FieldName: 'email',
        },
        {
          AttemptedValue: 'password',
          ErrorMessage: 'Invalid Password',
          ErrorType: 'ValidationError',
          FieldName: 'password',
        },
      ]

      const result = formatShiftErrors(errors)

      expect(result).toHaveLength(2)
      expect(result).toMatchSnapshot()
    })
  })
})
