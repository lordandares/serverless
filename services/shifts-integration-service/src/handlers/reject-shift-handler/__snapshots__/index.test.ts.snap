// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handlers:reject-shift-handler helpers :: formatShiftErrors should format error messages 1`] = `
Array [
  Object {
    "attemptedValue": "invalid@email",
    "errorMessage": "Invalid Email Address",
    "errorType": "ValidationError",
    "field": "email",
  },
  Object {
    "attemptedValue": "password",
    "errorMessage": "Invalid Password",
    "errorType": "ValidationError",
    "field": "password",
  },
]
`;

exports[`handlers:reject-shift-handler should end a shift and set the status to rejected 1`] = `
Object {
  "errors": Array [
    Object {
      "AttemptedValue": "invalid@email",
      "ErrorMessage": "Invalid Email Address",
      "ErrorType": "ValidationError",
      "FieldName": "email",
    },
  ],
  "shift": Object {
    "_id": "5d1eb8a4fc13ae1f85000000",
    "end": Object {
      "time": 2018-06-15T14:00:00.000Z,
    },
    "status": "rejected",
    "user": "1234",
  },
  "user": Object {
    "_id": "5d1eb8a4fc13ae1f85000001",
  },
}
`;

exports[`handlers:reject-shift-handler should error if the \`shiftId\` is missing 1`] = `"The Resource \\"shift\\" with ID of \\"Unknown\\" could not be found"`;
